lib/api/BackbeatAPI.js

```diff
@@ -211,60 +217,129 @@ class BackbeatAPI {
+            .setSourceClient(log)
+            .getMetadata(entry, log, (err, res) => {
+                if (err) {
+                    this._logger.error('could not retrieve object metadata ' +
+                        'from source during replication retry', { error: err });
+                    return cb(err);
+                }
+                const bucket = entry.getBucket();
+                const objectKey = entry.getObjectKey();
+                const mdObj = JSON.parse(res.Body);
+                const queueEntry =
+                    new ObjectQueueEntry(bucket, objectKey, mdObj);
+                queueEntry.setSite(entry.getSite());
+                return cb(null, queueEntry);
+            });
+    }
+
     /**
      * Builds the failed CRR response.
      * @param {String} cursor - The Redis HSCAN cursor
-     * @param {Array} hashes - The collection of Redis hashes for the iteration
-     * @return {Object} - The response object
+     * @param {Array} keys - The collection of Redis keys for the iteration
+     * @param {Function} cb - The callback function
+     * @return {undefined}
      */
-    _getFailedCRRResponse(cursor, hashes) {
+    _getFailedCRRResponse(cursor, keys, cb) {
         const response = {
             IsTruncated: Number.parseInt(cursor, 10) !== 0,
             Versions: [],
         };
         if (response.IsTruncated) {
             response.NextMarker = Number.parseInt(cursor, 10);
         }
-        for (let i = 0; i < hashes.length; i += 2) {
-            const [bucket, key, versionId, site] = hashes[i].split(':');
-            const entry = hashes[i + 1];
-            const value = JSON.parse(JSON.parse(entry).value);
-            response.Versions.push({
-                Bucket: bucket,
-                Key: key,
-                VersionId: versionId,
-                StorageClass: site,
-                Size: value['content-length'],
-                LastModified: value['last-modified'],
-            });
-        }
-        return response;
+        return async.eachLimit(keys, 10, (key, next) =>
+            this._redisClient.batch([['get', key]], (err, res) => {
+                if (err) {
+                    return next(err);
+                }
+                const [cmdErr, role] = res[0];
+                if (cmdErr) {
+                    return next(cmdErr);
+                }
+                // The key did not exist.
+                if (role === null) {
+                    return next();
+                }
+                const entry = new ObjectFailureEntry(key, role);
+                return this._getObjectQueueEntry(entry, (err, queueEntry) => {
+                    if (err) {
+                        // If we cannot retrieve the source object, then we no
+                        // longer can consider it a failed operation. Delete the
+                        // key used to monitor the failure.
+                        return this._deleteFailedCRRKey(key, next);
```

Comment:
Do we still want to include the object in the GET response even if we can’t retry it due to a missing source object? My sense is that if the source object is deleted, we should just remove the key from Redis as it's no longer representing an existing object with a FAILED status. It depends on how you view the API's monitoring functionality—i.e. is it tracking _all_ previous failures or _retryable_ failures.

Reply:
If the version does not exist anymore, I think there's no point in tracking it anymore, so indeed we should remove it from redis on getting a 404 from cloudserver IMHO.